File: client/modules/scanner-module/android/src/main/java/expo/modules/scannermodule/ScannerModule.kt
package expo.modules.scannermodule

import android.Manifest
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ScannerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ScannerModule")

    Events("onScanSuccess", "onError")

    AsyncFunction("requestCameraPermissionAsync") { promise ->
      val context = appContext.reactContext
      if (context == null) {
        promise.resolve(false)
        return@AsyncFunction
      }
      val result = androidx.core.content.ContextCompat.checkSelfPermission(
        context,
        Manifest.permission.CAMERA
      )
      if (result == android.content.pm.PackageManager.PERMISSION_GRANTED) {
        promise.resolve(true)
        return@AsyncFunction
      }
      val permissions = appContext.legacyModule<expo.modules.interfaces.permissions.Permissions>()
      if (permissions == null) {
        promise.reject("PERMISSIONS_NOT_AVAILABLE", "Permissions module not available")
        return@AsyncFunction
      }
      permissions.askForPermissionsAsync(
        object : expo.modules.core.Promise {
          override fun resolve(res: Any?) {
            val response = res as? Map<String, Any>
            val granted = response?.get(Manifest.permission.CAMERA) as? String
            promise.resolve(granted == "granted")
          }
          override fun reject(code: String, message: String?, throwable: Throwable?) {
            promise.reject(code, message, throwable)
          }
        },
        Manifest.permission.CAMERA
      )
    }

    View(ScannerModuleView::class) {
      Prop("isScanning") { view: ScannerModuleView, scanning: Boolean ->
        view.setIsScanning(scanning)
      }
      Events("onScanSuccess", "onError")
    }
  }
}

File: apps/client/modules/scanner-module/ios/ScannerModule.swift
import ExpoModulesCore
import AVFoundation

public class ScannerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ScannerModule")

    Events("onScanSuccess", "onError")

    AsyncFunction("requestCameraPermissionAsync") { () -> Bool in
      let status = AVCaptureDevice.authorizationStatus(for: .video)
      switch status {
      case .authorized:
        return true
      case .notDetermined:
        return await AVCaptureDevice.requestAccess(for: .video)
      default:
        return false
      }
    }

    View(ScannerModuleView.self) {
      Prop("isScanning") { (view: ScannerModuleView, scanning: Bool) in
        view.setScanning(scanning)
      }
      Events("onScanSuccess", "onError")
    }
  }
}
